cubedash/testutils/database.py

```diff
@@ -19,129 +20,252 @@
-                "db_port": host_port,
-                "db_database": "explorer_test",
-                "db_password": "badpassword",
-                "index_driver": "default",
-            }
+            db_url = f"postgresql://explorer_test_user:explorer_test_password@127.0.0.1:{host_port}/explorer_test_db"
+            logger.info("Docker container started", db_url=db_url)
+            yield db_url
         finally:
+            logger.info("Removing PostGIS container")
             container.remove(v=True, force=True)
 
 
 @pytest.fixture(scope="module")
-def odc_db(postgresql_server, tmp_path_factory, request):
+def odc_db(
+    postgresql_server: str,
+    env_name: str,
+    tmp_path_factory: pytest.TempPathFactory,
+    request: pytest.FixtureRequest,
+    worker_id,
+    testrun_uid,
+) -> Generator[str | ODCEnvironment]:
+    """Create an ODC Database per xdist worker and per pytest module
+
+    Write out an ODC Configuration File and set an Environment Variable specifying which ODC Env to Use.
+
+    We expose the current database by either writing out to a file (or, update a swath of environment variables)
+    rather than having a fixture which works at the Python level (SQLAlchemy engine or ODC Index/Database object)
+    because many of the integration tests work by launching the command line tools, which while not starting up
+    a whole new Python process, don't share much of the existing state and load their ODC Connection parameters
+    either from a file or environment variables.
+    """
     if postgresql_server == GET_DB_FROM_ENV:
-        yield None  # os.environ["DATACUBE_DB_URL"]
+        yield ODCConfig()[env_name]
     else:
-        postgres_url = "postgresql://{db_username}:{db_password}@{db_hostname}:{db_port}/{db_database}".format(
-            **postgresql_server
+        log = logger.bind(
+            db_url=postgresql_server, worker_id=worker_id, testrun_uid=testrun_uid
         )
-
-        new_db_database = request.module.__name__.replace(".", "_")
+        test_database_name = f"{request.path.stem}_{env_name}"
         # Wait for PostgreSQL Server to start up
         while True:
             try:
-                conn = psycopg2.connect(postgres_url)
-                conn.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
+                conn = psycopg2.connect(postgresql_server)
+                # Prevent starting a transaction, since 'create database' cannot run inside a transaction
+                conn.set_session(autocommit=True)
+
+                # assert not test_database_name.endswith('postgis')
+                # if test_database_name.endswith('postgis'):
+                # import web_pdb; web_pdb.set_trace()
+
+                try:
+                    with conn.cursor() as cur:
+                        log.info("Creating database", db_name=test_database_name)
+                        cur.execute(
+                            sql.SQL("CREATE DATABASE {db_name}").format(
```

Comment:
I had a dinner break since my earlier comments and I still love the thought of faster testing, but this feels like a big chunk of dynamic stuff that will be really hard to inspect/reproduce when things go mysteriously wrong in the CI.

We can easily create `$(nproc)` test databases when we spin up the current database container (just add a for loop loop around the create database part of the `create_db.sh` script in commit e6e823bc24077aa7f975c0012cff2c9177cd6899), isn't there some place in `conftest.py` (or similar) where we can patch in the worker number into the database-part of the engine connection string?